### ISSUE TITLE
route middleware and error page are called multiply on error occurred 

### ISSUE THREAD

**Issue by mercs600** [nuxt/framework#6951]
### Environment

Nuxt project info:                                                                                                                   18:01:54

------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.13.2`
- Nuxt Version:     `2.15.8`
- Package Manager:  `yarn@1.22.10`
- Builder:          `webpack`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

https://github.com/mercs600/github-vkg67s
https://stackblitz.com/edit/github-vkg67s 

### Describe the bug

I noticed that route middlewares are called multiply times when some error occurred (eg. fetch 404). I don't know why. 
You can change this line https://github.com/mercs600/github-vkg67s/blob/main/composables/useHelloPlugin.ts#L8 and remove 404 from path - then it works correctly.
But when fetch failed because of 404 you can see multiple console logs. 
The same thing happens for error.vue page - I setup there console.log which should be called once, instead of this page is called twice. 

### Logs

```shell
should call once for ssr
calling from composable [Function (anonymous)]
calling from error page, should have $hello object [Function (anonymous)]
should call once for ssr
calling from composable [Function (anonymous)]
calling from error page, should have $hello object [Function (anonymous)]
```

**Comment by Aareksio** [comment]
The reproduction is far from minimal and is hard to follow. You also shouldn't be using `router.onError()` inside route middleware which gets called with each navigation. I think your issue may have the root in #6957, may be something else though. I did not manage to have the function called twice on StackBlitz in the first place:

```
Nuxt CLI v3.0.0-rc.8                                                                                                                                                                      20:40:53
                                                                                                                                                                                          20:41:01
  > Local:    http://localhost:3000/ 
  > Network:  http://192.168.1.104:3000/

ℹ Vite server warmed up in 636ms                                                                                                                                                          20:41:05
✔ Vite server built in 1034ms                                                                                                                                                             20:41:05
✔ Nitro built in 439 ms                                                                                                                                                             nitro 20:41:06
ℹ Vite client warmed up in 3570ms                                                                                                                                                         20:41:07
should call once for ssr
calling from composable [Function (anonymous)]
[nitro] [dev] [unhandledRejection] TypeError: Failed to fetch
    at _0x4cbae9._fetch (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:15:413681)
    at _0x4cbae9.performHTTPCall (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:15:413246)
    at _0x4cbae9._write (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:15:407309)
    at _0x4cbae9.writeLatin1String (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:15:397824)
    at _0x95e122.writeLatin1String (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:15:217881)
    at https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:6:783677
    at writeGeneric (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:6:783958)
    at Socket._writeGeneric (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:6:237637)
    at Socket._write (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:6:237795)
    at doWrite (https://github-vkg67s.w.staticblitz.com/blitz.968a32de9d01484de0feb027dd69dcf18139a386.js:6:1012946)
```

---

Could you please attempt to recreate a minimal example -- just the code required for the issue to occur and nothing more?
